text/0000-edition-2021.md

```diff
@@ -0,0 +1,226 @@
+
+Having uniform behavior across editions has several benefits:
+
+* Reducing technical debt: the compiler is simpler if it has to consider fewer cases.
+* Minimizing cognitive overload: we want to avoid users having to think too much about what edition they are in. We would prefer that all Rust code feels the same, no matter what edition it is using, although we are willing to compromise on this principle if the benefits are large.
+    * Even when changes alter core parts of Rust, we are often able to introduce parts of those changes across all editions, which helps us to achieve more uniformity. For example, the `crate::foo::bar` paths introduced in Rust 2018 are also available on Rust 2015, since that syntax had no meaning before.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Edition cadence not specified
+
+This RFC explicitly does not attempt to specify the cadence for releasing editions. Historically, they have occurred every 3 years, if one considers Rust 1.0 (released in 2015) to be the "original edition". Our expectation is that after Rust 2021 is shipped, we will use the experiences from Rust 2018 and Rust 2021 to review how editions are working and potentially establish a cadence for future editions (or perhaps make other changes to the edition system). 
+
+There are various constraints on the timing of editions that have been identified over time:
+
+* Even if it's automated, migrating to a new edition can require a significant investment of time, particularly for production users with many crates. We don't wish to release Editions at a rate that makes it hard for such users to keep up.
+* At the same time, we don't wish to wait too long in between editions. We want to maintain the spirit of Rust's train model, so that folks don't feel a "rush" to get work done by any particular deadline.
+* Having a generally known time when editions will be released is helpful in planning feature development.
+* We do not wish to release an empty edition. Sometimes there simply won't be any backwards incompatible changes and thus no reason to issue an edition.
+
+## Definition: migration
+
+**Migrations** are the "breaking changes" introduced by the edition, except of course that since editions are opt-in, no code actually breaks. 
+
+## Definition: migration lint
+
+Migrations typically come with one or more **migration lints**. Each lint warns about code that will need to change in order to move to the new edition. The lints typically contain "suggestions", which is a diff that can be applied to make the code work in the new edition. In some cases, the rewrite may be too complex for the lint to make a suggestion. In addition, some suggestions are marked as "not machine applicable", if they are not known to be semantics preserving.
+
+## Tooling workflow
```

Comment:
I think a **requirement** for any edition change RFC that can potentially cause breakages would be:


 - List out all the potential breakages, big and small
 - Potentially declare some breakages as out of scope for the migration. For example, this can be stuff where too many macros are involved.
 - Have an in-depth design for a migration lint that will migrate code such that it compiles on both editions. The compiler team should be quickly consulted to ensure the lint is feasible. The migration lint must handle _all_ cases that are declared in scope
 - Optionally have an in-depth design for an idiom lint that will clean up the migrated code (potentially making it compile on the new edition only)


I think such a requirement should be in this RFC, since we [already have an RFC that is missing details on this](https://github.com/rust-lang/rfcs/pull/3090/files#r583971128).

Reply:
Hmm, I'm not sure if I agree with this, actually. I don't know that I think the migration lint needs to have an in-depth design, but I do think it makes sense to discuss the scenarios that will need migration. I think the rule of thumb wants to be that for each scenario we know:

* what kind of code users should write instead; and,
* whether we expect to be able to migrate it automatically (and, if so, enough details as to how we will do it that the team is convinced it will work).

I think that RFC #3067 is a good example of this. We decided to hold off on moving forward until we had a good plan around recursive macros, precisely because we couldn't answer the first question, even though there was general consensus that we wouldn't need automatic migration.

Reply:
This is basically the same standard I think RFCs should take for most things, to be honest. In-depth discussion of how things will work often works better in the PR and implementation phase, but we want to go deep enough to be reasonably confident the implementation is feasible and will succeed.

Reply:
I pushed some text, @Manishearth -- take a look.

Reply:
Yeah, this is fine by me. I think there needs to be a feasability check from the compiler team (2018 had this problem amongst others), but yes, essentially, we need to make sure we know what the before/afters are.

Could we perhaps go into detail on "describe how the code will be modified" and state that this will likely be by explaining the transformation done by one or more edition lints, plus potentially some idiom lints?

Reply:
When I say in-depth design I really mean just that^, that the before/after should be more clearly designed, not just "we'll have a migration" 😄 I perhaps stated it more strongly than I needed to!

Reply:
Added a few more notes to that effect.

Reply:
Sorry it took me a while to get back to this, but can we have a clear section that prescribes a set of rules for proposed edition changes? I'm having to repeatedly explain parts of this and would prefer if this were in one place.

Reply:
I'm happy to provide text for this if you would like!

Reply:
@Manishearth can you post the sample copy so I have a better idea what you mean?

Reply:
@nikomatsakis I wrote a doc on edition policies, and the sample copy is roughly the "Migration plan requirements" section: https://hackmd.io/@nbw0Mih0RLieoV6JBAtI3Q/SyFshP84O

Reply:
I've added changes, @Manishearth, to this effect. I adapted your language and lightly edited it.